diff --git a/core/modules/migrate/src/Plugin/migrate/source/SqlBase.php b/core/modules/migrate/src/Plugin/migrate/source/SqlBase.php
index e5c2b07591..dfb5e6cca3 100644
--- a/core/modules/migrate/src/Plugin/migrate/source/SqlBase.php
+++ b/core/modules/migrate/src/Plugin/migrate/source/SqlBase.php
@@ -425,3 +425,10 @@ protected function mapJoinable() {
-    if ($this->trackChanges) {
+    // However, only do this when we know we want to refresh already
+    // imported rows, otherwise we slow down initial imports too.
+    // (When migrating using the UI, the batch system is used. This means a
+    // migration runs across many requests, and not joining against the map
+    // table implies every batch request must iterate over every source row,
+    // causing a performance regression.
+    $acquia_migrate_unavailable = !class_exists('\Drupal\acquia_migrate\Batch\MigrationBatchManager');
+    if ($this->trackChanges && ($acquia_migrate_unavailable || \Drupal\acquia_migrate\Batch\MigrationBatchManager::$isRefreshing)) {
       return FALSE;
     }
